### PR TITLE
React on events

### PR DESCRIPTION
**Pull request by LiliC**
So far we were only reacting on `Habitat` and `Pod` events. This PR adds `Deployment` and `ConfigMap` events. This PR also stores the cache. This cache will be used in the subsequent PR and it will optimise the unnecessary calls to the API. My reasons for splitting is for easier review.

Couple of points:
- In the `handleHab` function I extracted the handling into separate functions to be consistent with the other resource handlers.
- Adding enqueuing in the `handlePodAdd` improves the time the IP is written into the `peer-watch-file` CM.
- I have explored reacting to `Secrets` as well and storing them. But we do not store any information about which `Secret` belongs to which `Habitat`. Therefor we cannot enqueue the correct `Habitat` object and would need to enqueue al of the existing `Habitat`s in all namespaces, which would just slow things down.
- We do react on `ConfigMap` events even if it's one `CM` for all the `Habitat`s. We do this because if there is an add/update/delete event on the peer-watch-file `CM` we do want to reconcile again asap no matter the `Habitat`.
- As for the `Pod`s as @asymmetric already [discovered](https://github.com/kinvolk/habitat-operator/issues/65#issuecomment-324927718) we cannot use the cache to get rid of the API calls.

Closes https://github.com/kinvolk/habitat-operator/issues/66.

### CODE REVIEW

File: pkg/habitat/controller/controller.go [hunk 1]
```diff
@@ -268,6 +448,7 @@ func (hc *HabitatController) handleConfigMap(h *crv1.Habitat) error {
 			if err := hc.writeLeaderIP(cm, ""); err != nil {
 				return err
 			}
+			return nil
```
Comment (LiliC) [inline pkg/habitat/controller/controller.go]: We return early here because we have finished with handling the CM for this event.

File: pkg/habitat/controller/controller.go [hunk 2]
```diff
@@ -120,94 +126,269 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
-		if !apierrors.IsAlreadyExists(err) {
-			return err
+func (hc *HabitatController) handleDeployAdd(obj interface{}) {
+	d, ok := obj.(*appsv1beta1.Deployment)
+	if !ok {
+		level.Error(hc.logger).Log("msg", "Failed to type assert deployment", "obj", obj)
+		return
+	}
+
+	if d.ObjectMeta.Labels["habitat"] == "true" {
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: How about adding a function like `func isHabitatObject(meta *<type of ObjectMeta>) bool` that checks the label?

File: pkg/habitat/controller/controller.go [hunk 3]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+	hc.habInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    hc.handleHabAdd,
+		UpdateFunc: hc.handleHabUpdate,
+		DeleteFunc: hc.handleHabDelete,
+	})
+}
+
+func (hc *HabitatController) cacheDeployment() {
+	source := cache.NewListWatchFromClient(
+		hc.config.KubernetesClientset.AppsV1beta1().RESTClient(),
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: My lack of client-go knowledge probably leaves me confused about the fact that you pass different things as `cache.Getter` for each watcher. It's `hc.config.HabitatClient` for watching Habitats, `hc.config.KubernetesClientset.AppsV1beta1().RESTClient()` for watching deployments and `hc.config.KubernetesClientset.CoreV1().RESTClient()` for config maps and pods. So for each of those things do you need a separate client or something?

File: pkg/habitat/controller/controller.go (hunk 2 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Also, use `crv1.HabitatLabel` instead of `"habitat"`.

File: pkg/habitat/controller/controller.go [hunk 4]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+		return
+	}
+
+	if d.ObjectMeta.Labels["habitat"] == "true" {
+		hc.enqueue(obj)
+	}
+}
+
+func (hc *HabitatController) handleCMAdd(obj interface{}) {
+	cm := obj.(*apiv1.ConfigMap)
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: In other handlers you check explicitly if the type assert succeeded. Why not doing it here?

Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Also, this code is duplicated in update and delete handlers - maybe just put all this in one function? Unless you have plans in the immediate future to change one of the config map handlers.

File: pkg/habitat/controller/controller.go [hunk 5]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+func (hc *HabitatController) handleCMDelete(obj interface{}) {
+	cm := obj.(*apiv1.ConfigMap)
+	if cm.ObjectMeta.Labels["habitat"] == "true" {
+		hc.enqueueNs(cm.GetNamespace())
+	}
+}
+
+func (hc *HabitatController) enqueueNs(ns string) {
+	cache.ListAll(hc.habInformer.GetStore(), labels.Everything(), func(obj interface{}) {
+		h := obj.(*crv1.Habitat)
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Explicit check for type assertion?

File: pkg/habitat/controller/controller.go [hunk 6]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+		if h.Namespace == ns {
+			hc.enqueue(h)
 		}
+	})
+}
 
-		d, err = hc.config.KubernetesClientset.AppsV1beta1Client.Deployments(h.Namespace).Get(deployment.Name, metav1.GetOptions{})
+func (hc *HabitatController) handlePodAdd(obj interface{}) {
+	pod := obj.(*apiv1.Pod)
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Explicit type assertion check?

File: pkg/habitat/controller/controller.go [hunk 7]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
 	}
 
-	return nil
+	newPod, ok2 := newObj.(*apiv1.Pod)
+	if !ok2 {
+		level.Error(hc.logger).Log("msg", "Failed to type assert pod", "obj", newObj)
+		return
+	}
+
+	if !hc.podNeedsUpdate(oldPod, newPod) {
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Shouldn't you check the habitat label on the new pod too before checking if it needs update?

File: pkg/habitat/controller/controller.go [hunk 8]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+	}
+
+	h, err := hc.getHabitatFromPod(newPod)
+	if err != nil {
+		if hErr, ok := err.(habitatNotFoundError); !ok {
+			level.Error(hc.logger).Log("msg", hErr)
+			return
+		}
+
+		// This only means the Pod and the Habitat watchers are not in sync.
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Is this harmless?

File: pkg/habitat/controller/controller.go [hunk 9]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+	}
+
+	h, err := hc.getHabitatFromPod(pod)
+	if err != nil {
+		if hErr, ok := err.(habitatNotFoundError); !ok {
+			level.Error(hc.logger).Log("msg", hErr)
+			return
+		}
+
+		// This only means the Pod and the Habitat watchers are not in sync.
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Ditto about being harmless.

File: pkg/habitat/controller/controller.go [hunk 10]
```diff
@@ -120,94 +126,268 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+	hc.cMInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    hc.handleCMAdd,
+		UpdateFunc: hc.handleCMUpdate,
+		DeleteFunc: hc.handleCMDelete,
+	})
 }
 
-func (hc *HabitatController) handleHabitatCreation(h *crv1.Habitat) error {
-	level.Debug(hc.logger).Log("function", "handleHabitatCreation", "msg", h.ObjectMeta.SelfLink)
+func (hc *HabitatController) watchPods(ctx context.Context) {
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: This function is similar to `cacheFoo` functions. Maybe you could make it consistent with the rest? You could:
- either change the `cacheFoo` functions to run the informer in the same function, so you don't need to store them as members of `HabitatController` (with `habInfomer` being an exception, because it seems to be used elsewhere)
- or change the `watchPods` function (and rename it to `cachePod`) to set the to-be-added `podInformer` member of `HabitatController` and run it in the `Run` function (just like you are doing with the `cacheFoo` functions.

File: pkg/habitat/controller/controller.go (hunk 1 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: We do return nil just a few lines below anyway. Did you add this to avoid the info about config map being created?

File: pkg/habitat/controller/controller.go [hunk 11]
```diff
@@ -656,8 +755,44 @@ func (hc *HabitatController) syncHabitat(key string) error {
...
+	}
+
+	// Create Deployment, if it doesn't already exist.
+	_, err = hc.config.KubernetesClientset.AppsV1beta1Client.Deployments(h.Namespace).Create(deployment)
+	if err != nil {
+		// Was the error due to the Deployment already existing?
+		if !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+		_, err = hc.config.KubernetesClientset.AppsV1beta1Client.Deployments(h.Namespace).Get(deployment.Name, metav1.GetOptions{})
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: What's the point in confirming that the deployment exists by trying to get it if the error already stated the the deployment already exists?

File: pkg/habitat/controller/controller.go [hunk 12]
```diff
@@ -709,6 +844,9 @@ func newConfigMap(ip string) *apiv1.ConfigMap {
 	return &apiv1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: configMapName,
+			Labels: map[string]string{
+				crv1.HabitatLabel: "true",
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: I'm sometimes wondering if we should make `"true"` also some constant, like `crv1.HabitatLabel` is now. Would it be an overkill?

File: pkg/habitat/controller/controller.go (hunk 1 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Yes exactly, otherwise the log msgs were confusing and it stated each time "created peer....". Returning earlier made sense here.

File: pkg/habitat/controller/controller.go (hunk 10 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Actually not really. As I mentioned in the description of the PR we do not actually cache the pods, we just react on events. So I thing the name `watchPods` is appropriate.

> ...so you don't need to store them as members of HabitatController

We will use the informer in the very next PR, as mentioned in the description. I split it up for easier review. This is just about reacting on events.

File: pkg/habitat/controller/controller.go (hunk 4 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: > Also, this code is duplicated in update and delete handlers - maybe just put all this in one function?

Not sure I understand exactly what you mean by that?

File: pkg/habitat/controller/controller.go (hunk 12 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: I think it's an overkill but I see your point. 👍

File: pkg/habitat/controller/controller.go (hunk 4 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: The body of the `handleCMAdd`, `handleCMUpdate` and `handleCMDelete` are all the same. So:
- either just make it one function and pass it to the `AddEventHandler` for `hc.cMInformer`,
- or (better IMO) just move this code to a separate function and call it from all the `handleCMFoo` functions.

Comment (krnowak) [reply pkg/habitat/controller/controller.go]: But don't do any of those if the followup PR will make the `handleCMFoo` functions all different.

File: pkg/habitat/controller/controller.go (hunk 1 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Maybe add some debug about clearing the config map then?

File: pkg/habitat/controller/controller.go (hunk 10 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Alright, let's keep it like that then.

File: pkg/habitat/controller/controller.go (hunk 7 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: This code was just taken from the already existing codebase. 

But after having a look at the code, I think we do not need to do the check in the first place. As in the `watchPods` function we already filter and react only on events of the pods with the `"habitat" : "true"` label. So will remove the first check as that condition would never happen.
https://github.com/kinvolk/habitat-operator/blob/a388c7738323daaf35126dc09e90bca180031030/pkg/habitat/controller/controller.go#L199

File: pkg/habitat/controller/controller.go (hunk 11 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: This is just moved from the old codebase.

But I agree with you, and I commented this in the PR that introduced this part of the code. My plan is to actually first do a get from the cache and then create a deployment if it does not exist. I already started implementing this in the part two of this PR, wdyt of that plan?

File: pkg/habitat/controller/controller.go (hunk 3 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: So each of the rest clients has their own registered resources, so for example the `AppsV1Beta1` knows about the `deployments` resource, etc. 

There are no validations for when we actually do [cache them](https://github.com/kinvolk/habitat-operator/blob/c541b040326c0e01de8d397289cd336c70a97025/vendor/k8s.io/client-go/tools/cache/listwatch.go#L61), so you can pass in any of clients, but if you actually try it out it will print out an error when doing the actual request. The error looks something like this:
```
ERROR: github.com/kinvolk/habitat-operator/pkg/habitat/controller/controller.go:116: Failed to list *v1beta1.Deployment: the server could not find the requested resource
```

tl;dr: We need to pass exactly those clients.

File: pkg/habitat/controller/controller.go (hunk 4 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Makes sense, extracted and moved the content of the `enqueueNS` into the `enqueueCM` function.

File: pkg/habitat/controller/controller.go (hunk 1 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Will do, although sometimes I feel like we log too much maybe. I personally only look at the logs that actually write the IP into the CM rather then the other logs, ie the ones that contain the IP.

File: pkg/habitat/controller/controller.go (hunk 9 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: This part of the code was just moved to a more sensible place, the comment was added in a previous PR not by me. But in general syncing happens on each resource event so they will get in the correct state at another point.

**Comment by LiliC** [comment]
@krnowak Think I addressed or answered all the comments you had, can you PTALA.

File: pkg/habitat/controller/controller.go (hunk 3 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Alright, thanks.

File: pkg/habitat/controller/controller.go [hunk 13]
```diff
@@ -120,94 +126,262 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+
+func (hc *HabitatController) handlePodDelete(obj interface{}) {
+	pod, ok := obj.(*apiv1.Pod)
+	if !ok {
+		level.Error(hc.logger).Log("msg", "Failed to type assert pod", "obj", obj)
+		return
 	}
 
-	return nil
+	if isHabitatObject(&pod.ObjectMeta) == false {
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: `if !isHabitatObject(…)`.

File: pkg/habitat/controller/controller.go [hunk 14]
```diff
@@ -623,7 +718,7 @@ func (hc *HabitatController) processNextItem() bool {
 	}
 	defer hc.queue.Done(key)
 
-	err := hc.syncHabitat(key.(string))
+	err := hc.conform(key.(string))
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Maybe also add a comment about this panicking type assertion?

File: pkg/habitat/controller/controller.go [hunk 15]
```diff
@@ -723,3 +856,11 @@ func habitatKeyFromPod(pod *apiv1.Pod) string {
 
 	return key
 }
+
+func isHabitatObject(objMeta *metav1.ObjectMeta) bool {
+	if objMeta.Labels[crv1.HabitatLabel] != "true" {
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: Just `return objMeta.Labels[crv1.HabitatLabel] == "true"` should be fine.

File: pkg/habitat/controller/controller.go (hunk 1 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Not sure if there is such a thing like "too many logs", but rather "misleading/wrong logs". But sure, more logging can be added when a need arises.

File: pkg/habitat/controller/controller.go (hunk 11 above)
Comment (krnowak) [reply pkg/habitat/controller/controller.go]: Sure, sounds fine.

File: pkg/habitat/controller/controller.go [hunk 16]
```diff
@@ -120,94 +126,262 @@ func (hc *HabitatController) Run(ctx context.Context) error {
...
+func (hc *HabitatController) enqueueCM(obj interface{}) {
+	cm, ok := obj.(*apiv1.ConfigMap)
+	if !ok {
+		level.Error(hc.logger).Log("msg", "Failed to type assert ConfigMap", "obj", obj)
+		return
+	}
+
+	if isHabitatObject(&cm.ObjectMeta) == true {
+		cache.ListAll(hc.habInformer.GetStore(), labels.Everything(), func(obj interface{}) {
+			h := obj.(*crv1.Habitat)
```
[the remark this answers is not in the document]
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Actually I just did a check and put it in code rather then a comment.

File: pkg/habitat/controller/controller.go (hunk 14 above)
Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Added a check here as well.

**Comment by LiliC** [comment]
@krnowak Addressed your suggestions, can you have look again, thanks!

File: pkg/habitat/controller/controller.go [hunk 17]
```diff
@@ -268,6 +449,8 @@ func (hc *HabitatController) handleConfigMap(h *crv1.Habitat) error {
 			if err := hc.writeLeaderIP(cm, ""); err != nil {
 				return err
 			}
+			return nil
+			level.Info(hc.logger).Log("msg", "removed peer IP from ConfigMap", "name", newCM.Name)
```
Comment (krnowak) [inline pkg/habitat/controller/controller.go]: This is dead code. It should be moved above `return nil`.

Comment (LiliC) [reply pkg/habitat/controller/controller.go]: Oh god, wish there was a facepalm emoji on here. :/ Fixed.

**Comment by krnowak** [comment]
LFAD.
